pkg/controller/master/addon/patch_grafana.go

```
@@ -0,0 +1,175 @@
+	if aObj == nil || aObj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if aObj.Name != harvesterutil.RancherMonitoringName || aObj.Namespace != harvesterutil.CattleMonitoringSystemNamespaceName {
+		return aObj, nil
+	}
+
+	// deployed successfully
+	if aObj.Spec.Enabled && aObj.Status.Status == harvesterv1.AddonDeployed {
+		// patch pv ReclaimPolicy
+		_, pvname, err := h.patchGrafanaPVReclaimPolicy(harvesterutil.CattleMonitoringSystemNamespaceName, harvesterutil.GrafanaPVCName)
+
+		if err != nil {
+			return aObj, err
+		}
+
+		// annotate pv name
+		if pvname != "" {
+			return h.annotateGrafanaPVName(pvname, aObj)
+		}
+	}
+
+	// disabled successfully, the pv has already from Bound to Released/Available
+	if !aObj.Spec.Enabled && aObj.Status.Status == harvesterv1.AddonInitState {
+		logrus.Debugf("start to patch grafana pv ClaimRef")
+		return h.patchGrafanaPVClaimRef(aObj)
+	}
+
+	return aObj, nil
+}
+
+// return:
+// bool: really patched;  string: pv-name;    error
+func (h *Handler) patchGrafanaPVReclaimPolicy(namespace, name string) (bool, string, error) {
+	pvc, err := h.pvc.Cache().Get(namespace, name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// grafana may not enable pvc
+			logrus.Debugf("grafana PVC %s/%s is not found, skip patch ReclaimPolicy", namespace, name)
+			return false, "", nil
+		}
+
+		return false, "", fmt.Errorf("error fetching grafana pvc %s/%s: %v", namespace, name, err)
+	}
+
+	// pv name format: volumeName: pvc-a95cdf6b-26bc-4ed6-8aaf-5f5d6b0b6dbd
+	pv, err := h.pv.Cache().Get(pvc.Spec.VolumeName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Debugf("grafana PV %s is not found, retry", pvc.Spec.VolumeName)
+			return false, "", fmt.Errorf("grafana PV %s is not found, retry", pvc.Spec.VolumeName)
+		}
+
+		return false, "", fmt.Errorf("error fetching grafana pv %s: %v", pvc.Spec.VolumeName, err)
```

we could return the error directly since it already contains the proper error message.